### ISSUE TITLE
Example of vertx pac4j LDAP implementation

### ISSUE THREAD

**Issue by kpnarayanan** [pac4j/vertx-pac4j#68]
Hello all,

I have a LDAP server that I am trying to connect using a backend written in vertx. I am looking for an example / client program in vertx-pac4j which I can use to connect to the LDAP server. I browsed through the documentation available in Github but I was not able to locate the example.

May I know is there an example in the github to which I can refer. If not, can somebody kindly suggest any pointers  / source code using which I can write the LDAP client program specific to vertx-pac4j.

Regards,
Krishnaprasad

**Comment by millross** [comment]
I don’t think there’s anything vertx-specific about this. If you can configure a Pac4j client to do the LDAP authentication, and configure that as if it were any other pac4j client (as per the vertx-pac4j-demo) then it should work. 

vertx-pac4j-demo shows how to set up a range of different clients and deploy them into vert.x. 

I’m sure Jérôme can describe how to configure a pac4j client to use LDAP, and once you have that you should be able to just use that client directly in vertx-pac4j as shown in the demo.

Note that because pac4j clients are generally blocking, the vertx-pac4j implementation just wraps the auth call in executeBlocking (rather than being truly non-blocking).

**Comment by kpnarayanan** [comment]
Thanks for your reply. I went through the vertx-pac4j demo however I did not find anything related to pac4j-ldap. I got it working by using the source code from the section "Binding" in this webpage - http://www.ldaptive.org/.

**Comment by millross** [comment]
Why would there need to be something specific for ldap in the vertx-pac4j-demo?

**Comment by millross** [comment]
Please could you open up a discussion on the pac4j-users google group about why you needed to add the Ldaptive client into the mix? It seems to me that if there's a standard pac4j client missing, then that should be called out - it should be possible to use LDAP via pac4j already so I'm a little concerned that it sounds like it isn't.

**Comment by millross** [comment]
(I've just read your note on the vertx group, hence the concern)

**Comment by kpnarayanan** [comment]
I started with Ldaptive client first and then used vertx-pac4j which is the reason I mentioned both of them. I found that there are no missing client libraries for pac4j. I just provided the information that I am able to validate / authenticate using the two clients but not using vertx shiro.

**Comment by millross** [comment]
That's great news, thankyou :) I'm active on both groups by the way (my main interest is in vert.x). Score one for vertx-pac4j over vertx's own-built auth frameworks.

**Comment by leleuj** [comment]
I'm late on this, but indeed the LDAP support is provided by the `pac4j-ldap` module which itself heavily relies on the Ldaptive library: http://www.ldaptive.org/

A good example to start from: https://github.com/pac4j/pac4j/blob/master/pac4j-ldap/src/test/java/org/pac4j/ldap/test/tools/AuthenticatorGenerator.java#L19

**Comment by millross** [comment]
I think the main thing is that for OIDC we should offer the capability for global logout. It feels like as much of this as possible should be in core pac4j (e.g. the logout url for OIDC), and the capability to define whether or not global logout is supported (maybe via Optional return value for logout url). Of course it does raise questions about what happens in the event of being logged in on multiple clients. It might even be that the clients can support that generalised logout case by throwing a redirect action to the relevant url, which the implementation then just needs to follow.

**Comment by leleuj** [comment]
Yes, the global logout should be in core pac4j. For OIDC sure, but for CAS, SAML, Facebook... I think most identity providers require a centralized logout.

Every client should have something like a `logout` method to return the appropriate url to call for a centralized logout, which becomes a little harder with multiple clients. I planned that for 2.0 (https://github.com/pac4j/pac4j/issues/54), but I have the feeling the issue is raising, it could be done earlier, it's an addition to the main API, so not a breaking change.

**Comment by kpnarayanan** [comment]
I am not sure whether it would be right further discussing in this forum. The issue which I had raised in vertx google forum about the problem in shiro, may I know where should I post the question so that the issue is resolved

**Comment by millross** [comment]
The original vertx issue you raised using the vertx-shiro stuff without pac4j is in the right place. Vertx shirt auth is - I believe - provided by the vertx team.

**Comment by kpnarayanan** [comment]
The concern that I have is I have not got any response  / reply to the question that I had posted. I think I might have to wait a little longer.
